### ISSUE TITLE
Encoding of body and headers

### ISSUE THREAD

**Issue by howardcox** [php-mime-mail-parser/php-mime-mail-parser#179]
We frequently come across problems with remote mail servers sending us poorly encoded emails.

Examples include:
* Headers with non-ascii characters in them
* Body which does not match the charset header
* Body which is _mostly_ the stated charset, but with some manual bits which

Presently we fix this ourselves by normalising the encoding we get out of `php-mime-mail-parser` functions into UTF-8 (or in theory whatever site encoding we have set, but it's always UTF-8 for us). Of course, we run into problems if one of our developers forgets to fix the encoding of a value.

I wonder whether you would welcome a patch which would do this normalising in `php-mime-mail-parser` so that methods like `getHeader()` and `getMessageBody()` return strings in a consistent encoding, rather than simply returning exactly what was in the email.

There are a number of ways I can imagine controlling that behaviour, because not everyone will want it, but the most obvious would probably be an additional parameter on `setText()` which could store the setting. Alternatively parameters on the various get-methods could work, but I'd rather set it once than continually pass the encoding I want to use.

**Comment by eXorus** [comment]
Could you give me an example because for me we already do that.

getHeaders will return the value in UTF8
getHeadersRaw will return the value without encoding

**Comment by howardcox** [comment]
Here are two example of the kinds of thing we come across:

A body reported as us-acii but actually containing UTF-8 or more often Windows-1250 characters.

[Latin1Characters.txt](https://github.com/php-mime-mail-parser/php-mime-mail-parser/files/1708109/Latin1Characters.txt)

`mb_detect_encoding()` reports that `getMessageBody()` does return a UTF-8 string for this, *but* on inspection the special characters aren't right, and the fixes we apply corrects the encoding to be usable.


And headers with special characters in, usually From or To, sometimes subject:
[BadEncodedFromHeader.txt](https://github.com/php-mime-mail-parser/php-mime-mail-parser/files/1708149/BadEncodedFromHeader.txt)

In this case, the From header, `getHeader('from')` and `getAddresses('from')` (which is what we're actually using) return strings for which `mb_detect_encoding()` returns `false` for, unable to get the encoding, and when trying to use this as a UTF-8 string (for example postgres treats it as an invalid byte sequence).

**Comment by eXorus** [comment]
Thanks for your feedback.

For the BadEncodedFromHeader.txt I will agree to fix it in the library but do you have the fix ?

For the Latin1Characters.txt it's controversial because the message is wrong encoded as you said.

As you can see in this code:

```php
    public function decodeCharset($encodedString, $charset)
    {
        if (strtolower($charset) == 'utf-8' || strtolower($charset) == 'us-ascii') {
            return $encodedString;
        } else {
            return iconv($this->getCharsetAlias($charset), 'UTF-8//TRANSLIT//IGNORE', $encodedString);
        }
    }
```
We always want to return an utf8 string or us-ascii.

**Comment by fijiwebdesign** [comment]
This is a good example where a middleware can be used. https://github.com/php-mime-mail-parser/php-mime-mail-parser/pull/180

**Comment by eXorus** [comment]
@fijiwebdesign Yes for the second issue.

For the first issue I think we can fix it in the lib. 
But it's also possible to say remove all the existing code with charset management and put it in a middleware to give only UTF8 or us-ascii value.

**Comment by eXorus** [comment]
@howardcox Could you provide the fix for the first issue ?

For the second issue, you have now a middleware to fix it in the 2.10.0 version

**Comment by eXorus** [comment]
I can't reproduce anymore the both issues, maybe it has already been fixed.

```php
$Parser = new Parser();
$Parser->setText(file_get_contents('Latin1Characters.txt'));

$this->assertStringContainsString("Example subject", $Parser->getHeader("subject"));
$this->assertStringContainsString("My gr‡ndm‡ tıld yÚ˚r gr‡ndm‡", $Parser->getTextRaw());
$this->assertStringContainsString("My gr‡ndm‡ tıld yÚ˚r gr‡ndm‡", $Parser->getText());
```


```php
$Parser = new Parser();
$Parser->setText(file_get_contents('BadEncodedFromHeader.txt'));

$this->assertStringContainsString("Example subject", $Parser->getHeader("subject"));
$this->assertStringContainsString("Example ± Sender <example@example.com>", $Parser->getHeader('from'));
$this->assertEquals([[
    'display' => 'Example ± Sender',
    'address' => 'example@example.com',
    'is_group' => false,
    ]], $Parser->getAddresses('from'));
    }

```
